d2/src/main/java/com/linkedin/d2/balancer/util/WarmUpLoadBalancer.java

```diff
@@ -288,4 +288,22 @@ public TransportClient getClient(Request request, RequestContext requestContext)
     _usedServices.add(serviceName);
     return client;
   }
+
+  @Override
+  public void getClient(Request request, RequestContext requestContext, Callback<TransportClient> clientCallback)
```

Comment:
I see we are enabling the async path by overriding the default implementation here. Does this mean that we already took care of IC Handling for those executions that are done inside PropertyEventBus? If not, this will break those services that are reliant on IC to propagate some contexts

Reply:
My understanding is that once service properties are loaded in memory, we do not need to switch thread to read from d2, only if d2 is not loaded for the first time, this is a problem.

I remember when you debugged an issue last time, without warmup, the first ever call will always fail, but all the following calls pass, so I guess doing the async getClient call isn't a problem for all the existing load balancers, it's only a problem when d2 is not loaded in memory.

I don't think this is making the current situation worse. For the warm up load balancer, if it can load d2 properties in memory in advance, this getClient call doesn't need to be blocking. But I agree it will be a problem if warmup did not finish, then the 1st call will fail, but subsequent calls should work.

Reply:
That is correct, once the properties are loaded into memory this issue won't happen because there is no asynchronous activity involved and the call gets executed in the same calling thread.

If we are depending on Warmup to fix the IC issue, how do we handle a new host deployed to the fleet? How do we handle a new service being added? 

We should look at the IC problem more holistically to make it future proof IMHO. How can we make sure that all the asynchronous executions inside the load balancer are IC friendly so that we can support a true asynchronous load balancer without having to make assumptions?

Reply:
Is it already a problem today for those mps who do not enable warmUp? Will there be a problem when a new host is deployed? If so, I think we should probably put the correct IC restore in place.

I'm surprised that we're depending on warmup load balancer by leveraging the blocking call to make sure IC works. I'm also surprised that the call is never truly async since warmup is globally enabled.

Reply:
> Is it already a problem today for those mps who do not enable warmUp

I dont think any of the Load balancer out there is Asynchrounous since WarmUp is enabled globally and all the D2 calls are effectively is a Sync Call. 

> Will there be a problem when a new host is deployed

This is a danger in trusing Warmup to bring the state to memory to make the IC work. Warmup only works with the previously cached data on disk. For a new host there are no past d2 meta data information to pre-load the load balancer state